FEXHeaderUtils/FEXHeaderUtils/ScopedSignalMask.h

```diff
@@ -12,7 +12,13 @@
 
 namespace FHU {
   /**
-   * @brief A class that masks signals and locks a mutex until it goes out of scope. It is NOT SAFE to move across threads.
+   * @brief A drop-in replacement for std::lock_guard that masks POSIX signals while the mutex is locked
+   *
+   * Use this class instead of plain C++ locks when a signal handler may use
```

Comment:
The problem is not isolated in longjumps - it affects any lock that may be taken again from a signal handler, which is most locks for us

Reply:
> The problem is not related to longjumps

I thought one of the possible problems was that a signal handler may longjmp out of a scope that held a mutex, after which that mutex will not be acquirable anymore. Did I misunderstand that this happens?

> it affects any lock that may be taken again from a signal handler, which is most locks for us

Fair enough. So the signal handler interrupts a scope that held a mutex, and hence can't acquire that mutex.

> I don't think we can make a docstring that really explains all of the gotchas.

The gotcha really just boils down to "Mutexes stay locked when a POSIX signal handler is invoked, so it can't safely be locked again without precautions", doesn't it?

> We don't do this everywhere for performance reasons, and there's also a possibility of a guest deadlock if a syscall gets called while the signals are blocked

Those details are fine to omit in the docstring, I'd say. We don't need a 100% comprehensive essay about this class, but we do need to give the reader any clue whatsoever why the class exists in the first place (which was missing before this PR).

Reply:
> I thought one of the possible problems was that a signal handler may longjmp out of a scope that held a mutex, after which that mutex will not be acquirable anymore. Did I misunderstand that this happen

I haven't though about the longjump case, however the most common case is when a lock gets interrupted by a signal. We have to compile code / call syscalls / do everything the guest asks us in the signal handler to to pass the signal to the guest, which makes almost all of our codebase signal re-entrant.

So, eg, without the signal masking, while we're tracking the VMA structures, a signal might happen, and the guest signal handler might do mmap, which will lead to a deadlock.

Reply:
I weakened the description to drop the "Use this if..." form and to only list examples instead. How is that?